threads.go

```diff
@@ -0,0 +1,357 @@
+	var err error
+	if debug {
+		err = setLogLevels(map[string]logger.Level{
+			"threads":     logger.DEBUG,
+			"threadstore": logger.DEBUG,
+		}, true)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	t := &threads{
+		host:        h,
+		blocks:      bs,
+		dagService:  ds,
+		ctx:         ctx,
+		cancel:      cancel,
+		Threadstore: ts,
+	}
+	t.service, err = newService(t)
+	if err != nil {
+		return nil, err
+	}
+
+	listener, err := gostream.Listen(h, IPELProtocol)
+	if err != nil {
+		return nil, err
+	}
+	rpc := grpc.NewServer()
+	go rpc.Serve(listener)
+	pb.RegisterThreadsServer(rpc, t.service)
+
+	return t, nil
+}
+
+func (t *threads) Close() (err error) {
+	var errs []error
+	weakClose := func(name string, c interface{}) {
+		if cl, ok := c.(io.Closer); ok {
+			if err = cl.Close(); err != nil {
+				errs = append(errs, fmt.Errorf("%s error: %s", name, err))
+			}
+		}
+	}
+
+	//weakClose("host", t.host) @todo: fix panic on close
+	weakClose("dagservice", t.dagService)
+	weakClose("threadstore", t.Threadstore)
+
+	if len(errs) > 0 {
+		return fmt.Errorf("failed while closing threads; err(s): %q", errs)
+	}
+
+	t.cancel()
+	return nil
+}
+
+func (t *threads) Host() host.Host {
+	return t.host
+}
+
+func (t *threads) DAGService() format.DAGService {
+	return t.dagService
+}
+
+func (t *threads) Add(ctx context.Context, body format.Node, opts ...tserv.AddOption) (l peer.ID, r thread.Record, err error) {
+	// Get or create a log for the new node
+	settings := tserv.AddOptions(opts...)
+	lg, err := t.getOrCreateOwnLog(settings.Thread)
+	if err != nil {
+		return
+	}
+
+	// Write a node locally
+	coded, err := t.createNode(ctx, body, lg, settings)
+	if err != nil {
+		return
+	}
+
+	// Push out the new record
+	err = t.service.push(ctx, coded, settings.Thread, lg.ID, settings)
+	if err != nil {
+		return
+	}
+
+	return lg.ID, coded, nil
+}
+
+func (t *threads) Put(ctx context.Context, rec thread.Record, opts ...tserv.PutOption) error {
+	// Get or create a log for the new rec
+	settings := tserv.PutOptions(opts...)
+	lg, err := t.getOrCreateLog(settings.Thread, settings.Log)
+	if err != nil {
+		return err
+	}
+
+	// Save the record locally
+	// Note: These get methods will return cached nodes.
+	block, err := rec.GetBlock(ctx, t.dagService)
+	if err != nil {
+		return err
+	}
+	event, ok := block.(*cbor.Event)
+	if !ok {
+		return fmt.Errorf("invalid event")
+	}
+	header, err := event.GetHeader(ctx, t.dagService, nil)
+	if err != nil {
+		return err
+	}
+	body, err := event.GetBody(ctx, t.dagService, nil)
+	if err != nil {
+		return err
+	}
+	err = t.dagService.AddMany(ctx, []format.Node{rec, event, header, body})
+	if err != nil {
+		return err
+	}
+
+	// Update head
+	t.SetHead(settings.Thread, lg.ID, rec.Cid())
+	return nil
+}
+
+func (t *threads) Pull(ctx context.Context, id thread.ID, lid peer.ID, offset cid.Cid, opts ...tserv.PullOption) ([]thread.Record, error) {
```

Comment:
Currently, this just pulls from the given log. However, we probably want to pull from the _whole_ thread. Also, we don't really care what the result is here... any new stuff should be delivered over some kind of channel.